[top_earlgrey,fpga] Add facilities and guide for debugging on FPGA with an ILA (#18652)
lowRISC/opentitan: doc/contributing/fpga/debugging_with_ila.md

> # Debugging on an FPGA with an Integrated Logic Analyzer (ILA)

Reviewer: You define the acronym in the paragraph after so I think you can skip it here:

```suggestion
# Debugging on an FPGA with an ILA
```

This document provides instructions on observing signals on an FPGA with an integrated logic analyzer (ILA) in the scope of the OpenTitan project.

## Background

> When one observes a problem with OpenTitan instantiated on an FPGA, we recommend first exploring software-based and simulation-based approaches to debug it.

Reviewer: ```suggestion
When you observe a problem with OpenTitan instantiated on an FPGA, we recommend first exploring software-based and simulation-based approaches to debug it.
```

Second, try to reproduce the problem in RTL simulation.
A gate-level simulation of the FPGA design is possible in theory, but it may not be practical when peripherals are involved in the problem and no cycle-accurate model of those peripherals is available.
Even if those approaches do not reveal the root cause itself, they often provide valuable insight into the involved internal processes.
If none of those approaches reveal the root cause, it may be necessary to observe FPGA-internal signals, which is the purpose of an ILA.

## Requirements

This guide was written for OpenTitan on a [NewAE CW310 board](./get_a_board.md#chipwhisperer-cw310) with a Kintex 7 FPGA on it.
You further need to have Xilinx Vivado installed; see [Install Vivado](/guides/getting_started/install_vivado) for the currently supported version and installation instructions.
Finally, you need a connection to the JTAG port of the FPGA.
This guide was written using [Xilinx's Platform Cable USB II](https://www.xilinx.com/products/boards-and-kits/hw-usb-ii-g.html), although the first generation Platform Cable USB or the SmartLynq Data Cable may also work.

> [Set the FPGA board up](/guides/getting_started/setup_fpga.html) and connect the JTAG plug of Platform Cable USB II to J18 on the CW310 and its USB plug to your workstation where Vivado is installed.

Reviewer: ```suggestion
[Set up the FPGA board](/guides/getting_started/setup_fpga.html) and connect the JTAG plug of Platform Cable USB II to J18 on the CW310 and its USB plug to your workstation where Vivado is installed.
```

The commands to create and configure an ILA core can be obtained from Vivado's GUI, although we have to tweak them to work with our flow, which uses Vivado *in-memory projects*.

Open Vivado's GUI (creating a project is not necessary), click on *Window* -> *IP Catalog*, and search for *ILA*.
Double-click on *ILA (Integrated Logic Analyzer)* to open the customization window.
Set *Component Name* to the name of the module in RTL (`ila_0` in our example above).
In the *General Options* tab:
- Keep *Monitor Type* to *Native*.
- Keep *Number of Probes* to 1 unless you have added additional probe ports (`probe1` etc.) in RTL.
- Select *Sample Data Depth* according to your needs; this determines how many samples can be saved per trigger action.
  4096 should be feasible unless you are probing many signals.
- Keep *Same Number of Comparators for All Probe Ports* ticked and *Number of Comparators* at 1 unless you know you need different settings.
- If you can define a trigger condition based on probe signals, leave *Trigger In Port* unticked.

>   Otherwise, i.e., if you need to trigger the ILA from RTL, tick it and connect the added `trig_in` and `trig_in_ack` ports (see diagram on the left of the GUI window with *Show disabled ports* unticked).

Reviewer: ```suggestion
  If you need to trigger the ILA from RTL or otherwise, tick it and connect the added `trig_in` and `trig_in_ack` ports (see diagram on the left of the GUI window with *Show disabled ports* unticked).
```

- If metastability of probed signals is a concern, add one or multiple *Input Pipe Stages*.

> - Tick *Capture Control*, which will allow you to filter samples for capturing based on probe inputs.

Reviewer: ```suggestion
- Tick *Capture Control*, which allows you to filter samples for capturing based on probe inputs.
```

Click *Program device*.
As *Bitstream file*, select the file identified at the end of the previous section.
As *Debug probes file*, navigate to the `synth-vivado` directory identified in the previous section, then go to `lowrisc_systems_chip_earlgrey_cw310_0.1.runs/impl_1` and select `debug_nets.ltx`.
Click *Program* and wait for programming to complete.

You should now see a window with one tab per ILA.
Each tab has a waveform viewer at the top and two setting windows at the bottom.
The left bottom window has two tabs.
In the *Settings* tab, useful settings are mainly the number of windows and the position of the trigger in the window.
In the *Status* tab, the ILA trigger can be armed, disarmed, and activated immediately.
The right bottom window also has two tabs; the first defines the trigger conditions and the second the capture conditions (both based on probe signals of the ILA).
Configure each ILA according to your needs and hit the 'play' button in the *Status* tab to arm its trigger.

Last but not least, you need to run a test in which the FPGA does not get reprogrammed (otherwise the ILA or at least its configuration is lost).

> To this end, modify `sw/host/opentitanlib/src/test_utils/init.rs` to not load a bitstream:

Reviewer 2: Technically this only applies to tests that have a custom host-side opentitanlib-linked test harness. For tests that just use opentitantool to drive them, you would need to comment out [this](https://cs.opensource.google/opentitan/opentitan/+/master:rules/opentitan_test.bzl;drc=bbf0ea10cdfb6681c094894ae00d0fe471c3edfc;l=200) line instead.

Author: Fixed, thanks!

```diff
@@ -138,9 +138,6 @@ impl InitializeTest {
         // Create the UART first to initialize the desired parameters.
         let _uart = self.bootstrap.options.uart_params.create(&transport)?;

-        // Load a bitstream.
-        Self::print_result("load_bitstream", self.load_bitstream.init(&transport))?;

> -

Reviewer: I think it's fine to have a patch like this in the guide. We can always add this flag later if people use this guide and things are unclear.

> -

Reviewer 2: yeah we could do that eventually, but since this is a fairly invasive debug approach, not sure its worth the effort to architect it at this time.
